src/transport_picoquic.cpp

```diff
@@ -0,0 +1,828 @@
+                                     bool isServerMode,
+                                     LogHandler& logger)
+    : logger(logger)
+      , isServerMode(isServerMode)
+      , stop(false)
+      , transportStatus(TransportStatus::Connecting)
+      , serverInfo(server)
+      , delegate(delegate)
+{
+  if (isServerMode && tcfg.tls_cert_filename == NULL) {
+    throw InvalidConfigException("Missing cert filename");
+  }
+  else if (tcfg.tls_cert_filename != NULL) {
+    (void)picoquic_config_set_option(&config, picoquic_option_CERT,
+                               tcfg.tls_cert_filename);
+
+    if (tcfg.tls_key_filename != NULL) {
+      (void)picoquic_config_set_option(&config, picoquic_option_KEY,
+                                 tcfg.tls_key_filename);
+    } else {
+      throw InvalidConfigException("Missing cert key filename");
+    }
+  }
+}
+
+PicoQuicTransport::~PicoQuicTransport()
+{
+  setStatus(TransportStatus::Shutdown);
+  shutdown();
+}
+
+void PicoQuicTransport::shutdown()
+{
+
+  if (stop) // Already stopped
+    return;
+
+  stop = true;
+
+  if (picoQuicThread.joinable()) {
+    logger.log(LogLevel::info, "Closing transport pico thread");
+    picoQuicThread.join();
+  }
+
+  cbNotifyQueue.stopWaiting();
+  if (cbNotifyThread.joinable()) {
+    logger.log(LogLevel::info, "Closing transport callback notifier thread");
+    cbNotifyThread.join();
+  }
+
+  logger.log(LogLevel::info, "done closing transport threads");
+
+}
+
+
+TransportStatus
+PicoQuicTransport::status()
+{
+  return transportStatus;
+}
+
+void PicoQuicTransport::setStatus(TransportStatus status) {
+  transportStatus = status;
+}
+
+StreamId PicoQuicTransport::createStream(
+  const TransportContextId& context_id,
+  bool use_reliable_transport)
+{
+  logger.log(LogLevel::debug, "App requests to create new stream");
+
+  const auto &iter = active_streams.find(context_id);
+  if (iter == active_streams.end()) {
+    logger.log(LogLevel::warn, "Invalid context id, cannot create stream");
+    return 0;
+  }
+
+  const auto &cnx_stream_iter = iter->second.find(0);
+  if (cnx_stream_iter == iter->second.end()) {
+    logger.log(LogLevel::warn,
+               "Missing primary connection zero stream, cannot create streams");
+    return 0;
+  }
+
+  if (isServerMode) // Server does not create streams, clients must create these
+    return 0;
+  else if (not use_reliable_transport)
+    return 0;
+
+  // Create stream will add stream to active_streams
+  PicoQuicTransport::StreamContext * stream_cnx = createStreamContext(
+      cnx_stream_iter->second.cnx,
+      next_stream_id);
+
+  TransportContextId tcid = context_id;
+  StreamId stream_id = next_stream_id;
+
+  cbNotifyQueue.push([&] () {
+    delegate.on_new_stream(tcid, stream_id);
+  });
+
+  // Set next stream ID
+  next_stream_id += 4; // Increment by 4, stream type is first 2 bits
+
+  return stream_cnx->stream_id;
+}
+
+TransportContextId
+PicoQuicTransport::start()
+{
+  TransportContextId cid = 0;
```

Comment:
This isn't even used until the very end, why not declare it down there?

Reply:
I'll move it to be more readable.